python/Scripts/mxvalidate.py

```diff
@@ -23,11 +23,12 @@ def main():
 
     if opts.stdlib:
         stdlib = mx.createDocument()
-        filePath = os.path.dirname(os.path.abspath(__file__))
-        searchPath = mx.FileSearchPath(os.path.join(filePath, '..', '..'))
-        searchPath.append(os.path.dirname(opts.inputFilename))
-        libraryFolders = [ "libraries" ]
-        mx.loadLibraries(libraryFolders, searchPath, stdlib)
+        libraryFolders = [ mx.FilePath(mx.getDefaultDataLibraryPath()) ]
+        try:
+            mx.loadLibraries(libraryFolders, mx.FileSearchPath(), stdlib)
```

Comment:
This is a good step forward, but it brings up an subtle point about our data library handling, which is that we currently separate the data search path (e.g. `--path`) from the set of data library folders that have been requested by the user (e.g. `--library`).  Combining these into a single "default data library path" requires us to work around the normal behavior of helper functions such as `loadLibraries`, and this could become a source of confusion for future developers.

For consistency with existing practice, perhaps we should instead provide a `getDefaultDataSearchPath` method in Python, which would provide a path just _above_ the packaged `libraries` folder, and we could then invoke `loadLibraries` in the standard way:

```
libraryFolders = [ "libraries" ]
loadLibraries(getDefaultSearchPath(), libraryFolders, stdlib)
```

I'd be interested in your thoughts on this, and my own sense is that we should strictly follow our existing conventions until we're ready to upgrade our data library approach across the entire codebase.

Reply:
I was thinking as part of this work to provide "convenience" to libraries to veer users away from having to know about search paths and "special" folder names which need to be looked up (and is not future-proof).

Perhaps instead of being "stuck" on `loadLibrarires()` how about encapsulating this inside a `loadStandardDefinitions(mx.LIBRARY_PATH, doc)`  function ?  Either passing a document as an argument or returning one. Then there is no previous API to "fit" this into. mx.LIBRARY_PATH = "libraries" could be exposed,
allowing for a user to append subfolder paths if they only only want a library subset to be loaded.

It feels like exposing `getDefaultDataSearchPath()` maintains the same (more?) complexity as there is no other data to find except definition libraries. Perhaps if more is added to the package then this can be added in ?

BTW, I've used `loadLibraries()` with and without empty search paths, and have leaned towards empty search paths to avoid accidently loading in a folder from the wrong path. i.e. specify where the roots are and consider search paths as optional (not required as in the example you show).

Reply:
@kwokcb I see where you're coming from, but it's not just `loadLibraries` that makes this distinction between data search paths and data library folders, it's really the entire shader generation system in MaterialX.  Both the MaterialX Viewer and Graph Editor take separate `--path` and `--library` arguments on the command line, and we support a dedicated `MATERIALX_SEARCH_PATH` environment variable that extends the data search path without affecting the data library folders.  Until we're ready to replace this system _globally_ with a new one, I believe we need our data library approach in MaterialX Python to mirror the one in the Viewer and Graph Editor, with `MATERIALX_SEARCH_PATH` having the same impact in MaterialX Python that it has in MaterialX C++.

Reply:
I'm was probably stuck on the word "data", but reading over the docs on this it makes sense.

How about we use `getDefaultDataSearchPath()` but also add in `getDefaultLibraryFolder()` to return `libraries as I'm still stuck on hiding `libraries`. Then it becomes:
```
libraryFolders = [ getDefaultLibraryFolder() ]
loadLibraries(getDefaultDataSearchPath(), libraryFolders, stdlib)
```
Then nothing is hard-coded and future-proof.